libraries/matrix/api/src/main/kotlin/io/element/android/libraries/matrix/api/auth/ClientAuthenticationObserver.kt

```diff
@@ -0,0 +1,14 @@
+/*
+ * Copyright 2024 New Vector Ltd.
+ *
+ * SPDX-License-Identifier: AGPL-3.0-only
+ * Please see LICENSE in the repository root for full details.
+ */
+
+package io.element.android.libraries.matrix.api.auth
+
+import io.element.android.libraries.matrix.api.MatrixClient
+
+fun interface ClientAuthenticationObserver {
```

Comment:
expose this as a `MatrixAuthenticationService.Listener` and maybe adds a way to add and remove listeners?

Reply:
I'm not sure how to design that, since we need:

- `RustMatrixAuthenticationService` to have access to it so it can call `onClientAuthenticationSucceeded`.
- `MatrixClientsHolder` to somehow be able to be notified from this listener.

But `MatrixClientsHolder` itself can't implement the listener, since its constructor holds a `MatrixAuthenticationService` and it would be a cyclic dependency: `Service -> Holder -> Service`. This wasn't fun to avoid...

Also, if I rename it to `Listener`, it would be weird if a `Listener` had listeners itself... And I'm not sure how useful it would be to be able to add/remove them when we only have one. If we need to make it more flexible we can iterate in the future, IMO.

Reply:
I propose this implementation:

[listenToNewMatrixClient.patch](https://github.com/user-attachments/files/17677680/listenToNewMatrixClient.patch)

Tested OK on my side, let me know what you think.

Note: the patch does not include on tests.